gem/optimise.py

```diff
@@ -413,3 +480,565 @@ def aggressive_unroll(expression):
+
+    nodes_to_pull = tuple([n for n, node_number in gem_int.items()
+                           if nodes[node_number].value() == 1])
+    return nodes_to_pull
+
+
+def cse(node, quad_ind, arg_ind_flat):
+    """
+    common subexpression elimination
+    :param node:
+    :param quad_ind:
+    :param arg_ind:
+    :return:
+    """
+    # do not expand quadrature terms all the way
+    expand_cse = expand_all_product(node, arg_ind_flat)
+    monos_cse = flatten_sum(expand_cse, arg_ind_flat)
+    # need a firt pass of monos to combine terms which have same nodes for all arg_ind
+    # this should be in a loop if >2 arg_ind
+    if len(arg_ind_flat) > 1:
+        optimal_factors = find_optimal_factors(monos_cse, arg_ind_flat)
+    else:
+        optimal_factors = list()
+    # pull out the optimal factors and form factor_list
+    factor_lists = list()
+    factor_dict = OrderedDict().fromkeys(optimal_factors)
+    for of in factor_dict:
+        factor_dict[of] = list()
+    for mono in monos_cse:
+        all_factors = [f for g in (0, 1, 2) + arg_ind_flat for f in mono[g]]
+        for of in optimal_factors:
+            if of in all_factors:
+                all_factors.remove(of)
+                factor_dict[of].append(all_factors)
+                break
+        else:
+            factor_lists.append(all_factors)
+    for of, factors in factor_dict.items():
+        # factors = list of lists representing sum of products
+        if len(factors) == 1:
+            # just one product
+            factor_lists.append([of] + factors[0])
+        else:
+            if len(arg_ind_flat) > 2:
+                # more argument indices to process
+                if len(set(of.free_indices) & set(arg_ind_flat)) != 1:
+                    raise AssertionError("this should not happen")
+                ind = (set(of.free_indices) & set(arg_ind_flat)).pop()
+                new_arg_ind_flat = tuple([i for i in arg_ind_flat if i != ind])
+                if len(arg_ind_flat) - len(new_arg_ind_flat) != 1:
+                    raise AssertionError("this should not happen")
+                factor_lists.append([of, cse(sumproduct_2_node(factors),
+                                             quad_ind, new_arg_ind_flat)])
+            else:
+                # remember to factorise factor list here
+                factor_lists.append([of, sumproduct_2_node(factorise(factors, quad_ind, arg_ind_flat))])
+    return sumproduct_2_node(factorise(factor_lists, quad_ind, arg_ind_flat))
+
+
+def gen_lex_sequence(items):
+    """
+    generate lexicographical order of permutations.
+    e.g. (i,j,k) -> [(i,), (i,j), (i,j,k), (j), (j,k), (k)]
+    :param items:
+    :return:
+    """
+    if not items:
+        return list()
+    result = gen_lex_sequence(items[1:])
+    return [(items[0],)] + [(items[0],) + x for x in result] + result
+
+
+@singledispatch
+def _reorder(node, self, indices):
+    """
+    Reorder Sum and Product to promote hoisting
```

Comment:
It looks like it's reassociating sum and product trees so that terms with fewer free indices appear first.  Is that right?